--- openjdk/jdk/src/share/lib/security/java.security
+++ openjdk/jdk/src/share/lib/security/java.security
@@ -149,7 +149,8 @@
                com.sun.org.apache.xml.internal.utils.,\
                com.sun.org.apache.xml.internal.security.,\
                com.sun.org.glassfish.,\
-               org.jcp.xml.dsig.internal.
+               org.jcp.xml.dsig.internal.,\
+               org.GNOME.Accessibility.
 
 #
 # List of comma-separated packages that start with or equal this string
