--- openjdk/jdk/make/common/Program.gmk.old	2008-08-04 17:38:04.000000000 +0200
+++ openjdk/jdk/make/common/Program.gmk	2008-08-04 19:36:33.000000000 +0200
@@ -85,6 +85,9 @@
 	endif
     endif
     ifeq ($(PLATFORM), linux)
+        ifneq (,$(filter $(ARCH), s390 s390x))
+	    LDFLAGS += -rdynamic
+	endif
 	LDFLAGS += -lz -Wl,-z -Wl,origin
 	LDFLAGS += -Wl,--allow-shlib-undefined
 	LDFLAGS += -Wl,-rpath -Wl,\$$ORIGIN/../lib/$(LIBARCH)/jli
