--- cacao/cacao/configure.ac.orig	2008-08-12 21:07:16.000000000 +0200
+++ cacao/cacao/configure.ac	2008-08-12 21:09:11.000000000 +0200
@@ -74,13 +74,18 @@
     JAVA_ARCH="m68k"
     ;;
 
-mips | mipsel )
+mips )
     ARCH_DIR="mips"
     ARCH_CFLAGS="-D__MIPS__"
-    dnl Is this correct for mipsel?
     JAVA_ARCH="mips"
     ;;
 
+mipsel )
+    ARCH_DIR="mips"
+    ARCH_CFLAGS="-D__MIPS__"
+    JAVA_ARCH="mipsel"
+    ;;
+
 powerpc )
     ARCH_DIR="powerpc"
     ARCH_CFLAGS="-m32 -D__POWERPC__"
--- cacao/cacao/configure.orig	2008-08-12 21:07:16.000000000 +0200
+++ cacao/cacao/configure	2008-08-12 21:09:59.000000000 +0200
@@ -2836,10 +2836,16 @@
     JAVA_ARCH="m68k"
     ;;
 
-mips | mipsel )
+mips )
     ARCH_DIR="mips"
     ARCH_CFLAGS="-D__MIPS__"
-        JAVA_ARCH="mips"
+    JAVA_ARCH="mips"
+    ;;
+
+mipsel )
+    ARCH_DIR="mips"
+    ARCH_CFLAGS="-D__MIPS__"
+    JAVA_ARCH="mipsel"
     ;;
 
 powerpc )
