--- cacao/cacao/src/native/jni.c~	2008-09-10 18:42:21.000000000 +0200
+++ cacao/cacao/src/native/jni.c	2009-03-02 22:33:23.000000000 +0100
@@ -4295,6 +4295,8 @@
 
 *******************************************************************************/
 
+#define JNI_VERSION_CACAO 0xCACA0000
+
 jint JNI_GetDefaultJavaVMInitArgs(void *vm_args)
 {
 	JavaVMInitArgs *_vm_args;
@@ -4310,6 +4312,7 @@
 
     case JNI_VERSION_1_2:
     case JNI_VERSION_1_4:
+    case JNI_VERSION_CACAO:
 		_vm_args->ignoreUnrecognized = JNI_FALSE;
 		_vm_args->options = NULL;
 		_vm_args->nOptions = 0;
