--- openjdk/hotspot/make/linux/makefiles/vm.make~	2011-04-19 16:13:48.000000000 +0200
+++ openjdk/hotspot/make/linux/makefiles/vm.make	2011-04-20 16:41:04.332368132 +0200
@@ -96,6 +96,11 @@
 CPPFLAGS += -DDERIVATIVE_ID="\"$(DERIVATIVE_ID)\""
 endif
 
+DEB_MULTIARCH := $(shell dpkg-architecture -qDEB_HOST_MULTIARCH 2>/dev/null || true)
+ifneq (,$(DEB_MULTIARCH))
+CPPFLAGS += -DDEB_MULTIARCH="\"$(DEB_MULTIARCH)\""
+endif
+
 ifdef DISTRIBUTION_ID
 CPPFLAGS += -DDISTRIBUTION_ID="\"$(DISTRIBUTION_ID)\""
 endif
--- openjdk/hotspot/src/os/linux/vm/os_linux.cpp~	2011-04-20 16:38:32.172144967 +0200
+++ openjdk/hotspot/src/os/linux/vm/os_linux.cpp	2011-04-20 16:40:19.152301850 +0200
@@ -357,7 +357,11 @@
  *        ...
  *        7: The default directories, normally /lib and /usr/lib.
  */
+#ifdef DEB_MULTIARCH
+#define DEFAULT_LIBPATH	"/usr/lib/" DEB_MULTIARCH "/jni" ":/lib/" DEB_MULTIARCH ":/usr/lib/" DEB_MULTIARCH ":/usr/lib/jni:/lib:/usr/lib"
+#else
 #define DEFAULT_LIBPATH	"/usr/lib/jni:/lib:/usr/lib"
+#endif
 
 #define EXTENSIONS_DIR  "/lib/ext"
 #define ENDORSED_DIR    "/lib/endorsed"
