--- openjdk/hotspot/src/os/linux/vm/os_linux.cpp~	2009-10-22 07:08:23.000000000 +0200
+++ openjdk/hotspot/src/os/linux/vm/os_linux.cpp	2009-10-22 07:13:14.000000000 +0200
@@ -281,11 +281,7 @@
  *        ...
  *        7: The default directories, normally /lib and /usr/lib.
  */
-#if defined(AMD64) || defined(_LP64) && (defined(SPARC) || defined(PPC) || defined(S390))
-#define DEFAULT_LIBPATH "/usr/lib64:/lib64:/lib:/usr/lib"
-#else
-#define DEFAULT_LIBPATH "/lib:/usr/lib"
-#endif
+#define DEFAULT_LIBPATH	"/usr/lib/jni:/lib:/usr/lib"
 
 #define EXTENSIONS_DIR  "/lib/ext"
 #define ENDORSED_DIR    "/lib/endorsed"
