--- openjdk/hotspot/build/linux/makefiles/gcc.make~	2007-11-24 16:52:59.077912783 +0100
+++ openjdk/hotspot/build/linux/makefiles/gcc.make	2007-11-24 17:10:26.373711002 +0100
@@ -69,7 +69,7 @@
 CFLAGS += -g
 
 ARCHFLAG = $(ARCHFLAG/$(BUILDARCH))
-ARCHFLAG/i486    = -m32 -march=i586
+ARCHFLAG/i486    = -m32 -march=i586 -mtune=generic
 ARCHFLAG/amd64   = -m64
 ARCHFLAG/ia64    =
 ARCHFLAG/sparc   = -m32 -mcpu=v9
