--- openjdk/jdk/make/common/Defs-linux.gmk~	2009-05-26 20:17:23.000000000 +0000
+++ openjdk/jdk/make/common/Defs-linux.gmk	2009-05-26 20:22:17.000000000 +0000
@@ -99,9 +99,9 @@
 endif
 
 CC_OPT/NONE     = 
-CC_OPT/LOWER    = -O2
-CC_OPT/HIGHER   = -O3
-CC_OPT/HIGHEST  = -O3
+CC_OPT/LOWER    = -O2 -fno-stack-protector -U_FORTIFY_SOURCE
+CC_OPT/HIGHER   = -O3 -fno-stack-protector -U_FORTIFY_SOURCE
+CC_OPT/HIGHEST  = -O3 -fno-stack-protector -U_FORTIFY_SOURCE
 
 CC_OPT          = $(CC_OPT/$(OPTIMIZATION_LEVEL))
 
