--- openjdk/hotspot/make/linux/makefiles/gcc.make~	2007-11-29 12:38:22.578339247 +0100
+++ openjdk/hotspot/make/linux/makefiles/gcc.make	2007-11-29 12:42:28.742052571 +0100
@@ -103,7 +103,7 @@
 CFLAGS_WARN/BYFILE = $(CFLAGS_WARN/$@)$(CFLAGS_WARN/DEFAULT$(CFLAGS_WARN/$@)) 
 
 # The flags to use for an Optimized g++ build
-OPT_CFLAGS += -O3
+OPT_CFLAGS += -O2
 
 # Hotspot uses very unstrict aliasing turn this optimization off
 OPT_CFLAGS += -fno-strict-aliasing
--- openjdk/corba/make/common/Defs-linux.gmk~	2007-11-29 12:38:22.578339247 +0100
+++ openjdk/corba/make/common/Defs-linux.gmk	2007-11-29 12:42:52.452410236 +0100
@@ -76,8 +76,8 @@
 #
 # Default optimization
 #
-CC_HIGHEST_OPT = -O3
-CC_HIGHER_OPT  = -O3
+CC_HIGHEST_OPT = -O2
+CC_HIGHER_OPT  = -O2
 CC_LOWER_OPT   = -O2
 CC_NO_OPT      =
 
--- openjdk/jdk/make/common/Defs-linux.gmk~	2007-11-29 12:38:22.578339247 +0100
+++ openjdk/jdk/make/common/Defs-linux.gmk	2007-11-29 12:43:13.992735167 +0100
@@ -86,8 +86,8 @@
 #
 # Default optimization
 #
-CC_HIGHEST_OPT = -O3
-CC_HIGHER_OPT  = -O3
+CC_HIGHEST_OPT = -O2
+CC_HIGHER_OPT  = -O2
 CC_LOWER_OPT   = -O2
 CC_NO_OPT      =
 
