--- openjdk/hotspot/make/linux/makefiles/gcc.make.orig	2008-12-02 14:56:38.000000000 +0100
+++ openjdk/hotspot/make/linux/makefiles/gcc.make	2008-12-02 14:58:56.000000000 +0100
@@ -67,6 +67,7 @@
 CFLAGS += -fno-exceptions
 CFLAGS += -D_REENTRANT
 CFLAGS += -fcheck-new
+CFLAGS += -fno-stack-protector
 
 # Always generate full debuginfo on Linux.  It'll be in a separate
 # debuginfo package when building RPMs.
--- openjdk/jdk/make/common/Defs-linux.gmk.orig	2008-12-02 14:56:41.000000000 +0100
+++ openjdk/jdk/make/common/Defs-linux.gmk	2008-12-02 15:00:34.000000000 +0100
@@ -134,6 +134,8 @@
   LDFLAGS_COMMON        += $(LDFLAGS_COMMON_$(ARCH))
 endif
 
+CFLAGS_REQUIRED         += -fno-stack-protector
+
 #
 # Selection of warning messages
 #
--- openjdk/corba/make/common/Defs-linux.gmk.orig	2008-12-02 14:56:41.000000000 +0100
+++ openjdk/corba/make/common/Defs-linux.gmk	2008-12-02 14:58:59.000000000 +0100
@@ -120,6 +120,8 @@
   LDFLAGS_COMMON        += $(LDFLAGS_COMMON_$(ARCH))
 endif
 
+CFLAGS_REQUIRED         += -fno-stack-protector
+
 # Add in platform specific optimizations for all opt levels
 CC_HIGHEST_OPT += $(_OPT_$(ARCH))
 CC_HIGHER_OPT  += $(_OPT_$(ARCH))
