--- openjdk/hotspot/make/linux/makefiles/gcc.make.orig	2008-12-02 14:56:38.000000000 +0100
+++ openjdk/hotspot/make/linux/makefiles/gcc.make	2008-12-02 14:58:56.000000000 +0100
@@ -67,6 +67,7 @@
 CFLAGS += -fno-exceptions
 CFLAGS += -D_REENTRANT
 CFLAGS += -fcheck-new
+CFLAGS += -fno-stack-protector
 
 # Always generate full debuginfo on Linux.  It'll be in a separate
 # debuginfo package when building RPMs.
--- openjdk/corba/make/common/Defs-linux.gmk~	2008-03-28 20:11:54.075568406 +0100
+++ openjdk/corba/make/common/Defs-linux.gmk	2008-03-28 20:13:51.537340290 +0100
@@ -107,6 +107,8 @@
   LDFLAGS_COMMON        += $(LDFLAGS_COMMON_$(ARCH))
 endif
 
+CFLAGS_REQUIRED         += -fno-stack-protector
+
 # Add in platform specific optimizations for all opt levels
 CC_HIGHEST_OPT += $(_OPT_$(ARCH))
 CC_HIGHER_OPT  += $(_OPT_$(ARCH))
--- openjdk/jdk/make/common/Defs-linux.gmk~	2008-03-28 20:11:54.095568707 +0100
+++ openjdk/jdk/make/common/Defs-linux.gmk	2008-03-28 20:14:15.037694787 +0100
@@ -117,6 +117,8 @@
   LDFLAGS_COMMON        += $(LDFLAGS_COMMON_$(ARCH))
 endif
 
+CFLAGS_REQUIRED         += -fno-stack-protector
+
 # Add in platform specific optimizations for all opt levels
 CC_HIGHEST_OPT += $(_OPT_$(ARCH))
 CC_HIGHER_OPT  += $(_OPT_$(ARCH))
