# unused patch

--- openjdk/hotspot/make/solaris/makefiles/gcc.make~	2009-10-02 23:16:39.000000000 +0200
+++ openjdk/hotspot/make/solaris/makefiles/gcc.make	2009-10-22 07:05:58.000000000 +0200
@@ -109,7 +109,7 @@
 
 
 # Compiler warnings are treated as errors 
-WARNINGS_ARE_ERRORS = -Werror 
+WARNINGS_ARE_ERRORS = -Werror -Wno-error=format
 # Enable these warnings. See 'info gcc' about details on these options
 ADDITIONAL_WARNINGS = -Wpointer-arith -Wconversion -Wsign-compare 
 CFLAGS_WARN/DEFAULT = $(WARNINGS_ARE_ERRORS) $(ADDITIONAL_WARNINGS) 
--- openjdk/hotspot/make/linux/makefiles/gcc.make~	2009-10-22 06:58:44.000000000 +0200
+++ openjdk/hotspot/make/linux/makefiles/gcc.make	2009-10-22 07:06:00.000000000 +0200
@@ -98,7 +98,7 @@
 endif
 
 # Compiler warnings are treated as errors
-WARNINGS_ARE_ERRORS = -Werror
+WARNINGS_ARE_ERRORS = -Werror -Wno-error=format
 
 # Except for a few acceptable ones
 # Since GCC 4.3, -Wconversion has changed its meanings to warn these implicit
