--- openjdk/hotspot/make/linux/makefiles/gcc.make~	2013-06-28 17:34:51.780388910 +0200
+++ openjdk/hotspot/make/linux/makefiles/gcc.make	2013-06-28 17:53:10.531781345 +0200
@@ -142,11 +142,6 @@
   CFLAGS += -pipe
 endif
 
-# Compiler warnings are treated as errors
-ifneq ($(COMPILER_WARNINGS_FATAL),false)
-WARNINGS_ARE_ERRORS = -Werror
-endif
-
 # Except for a few acceptable ones
 # Since GCC 4.3, -Wconversion has changed its meanings to warn these implicit
 # conversions which might affect the values. To avoid that, we need to turn
