--- openjdk/hotspot/make/solaris/makefiles/gcc.make~	2008-12-02 19:48:28.000000000 +0100
+++ openjdk/hotspot/make/solaris/makefiles/gcc.make	2008-12-02 19:50:49.000000000 +0100
@@ -110,7 +110,7 @@
 
 # Compiler warnings are treated as errors 
 # Commented out for now because of gcc 4.3 warnings OpenJDK isn't ready for
-# WARNINGS_ARE_ERRORS = -Werror 
+# WARNINGS_ARE_ERRORS = -Werror -Wno-error=format
 # Enable these warnings. See 'info gcc' about details on these options
 ADDITIONAL_WARNINGS = -Wpointer-arith -Wconversion -Wsign-compare 
 CFLAGS_WARN/DEFAULT = $(WARNINGS_ARE_ERRORS) $(ADDITIONAL_WARNINGS) 
--- openjdk/hotspot/make/linux/makefiles/gcc.make~	2008-12-02 19:48:51.000000000 +0100
+++ openjdk/hotspot/make/linux/makefiles/gcc.make	2008-12-02 19:51:04.000000000 +0100
@@ -99,7 +99,7 @@
 endif
 
 # Compiler warnings are treated as errors
-# WARNINGS_ARE_ERRORS = -Werror
+# WARNINGS_ARE_ERRORS = -Werror -Wno-error=format
 
 # Except for a few acceptable ones
 # Since GCC 4.3, -Wconversion has changed its meanings to warn these implicit
