--- openjdk/hotspot/src/os_cpu/linux_x86/vm/linux_x86_32.s~	2009-08-06 01:08:57.000000000 -0700
+++ openjdk/hotspot/src/os_cpu/linux_x86/vm/linux_x86_32.s	2009-08-06 01:09:27.000000000 -0700
@@ -650,3 +650,4 @@ _Atomic_cmpxchg_long:
         popl     %ebx
         ret
 
+.section        .note.GNU-stack, "", @progbits
--- openjdk/hotspot/src/os_cpu/linux_x86/vm/linux_x86_64.s~	2009-08-06 01:08:57.000000000 -0700
+++ openjdk/hotspot/src/os_cpu/linux_x86/vm/linux_x86_64.s	2009-08-06 01:09:27.000000000 -0700
@@ -400,3 +400,4 @@ _Atomic_cmpxchg_long:
         addq     $4,%rdx
         jg       4b
         ret
+.section        .note.GNU-stack, "", @progbits
