--- openjdk/jdk/src/solaris/classes/sun/awt/X11/XWindowPeer.java.orig
+++ openjdk/jdk/src/solaris/classes/sun/awt/X11/XWindowPeer.java
@@ -1143,6 +1143,7 @@
     boolean isOverrideRedirect() {
         return XWM.getWMID() == XWM.OPENLOOK_WM ||
             XWM.getWMID() == XWM.METACITY_WM ||
+            XWM.getWMID() == XWM.COMPIZ_WM ||
             Window.Type.POPUP.equals(getWindowType());
     }
 
