--- jsorig	2008-12-05 11:43:39.000000000 -0500
+++ openjdk/jdk/src/share/lib/security/java.security	2008-12-05 11:43:45.000000000 -0500
@@ -127,7 +127,7 @@
 # passed to checkPackageAccess unless the
 # corresponding RuntimePermission ("accessClassInPackage."+package) has
 # been granted.
-package.access=sun.,com.sun.xml.internal.ws.,com.sun.xml.internal.bind.
+package.access=sun.,com.sun.xml.internal.ws.,com.sun.xml.internal.bind.,org.GNOME.Accessibility.,org.GNOME.Bonobo.
 
 #
 # List of comma-separated packages that start with or equal this string
