--- openjdk/jdk/src/share/classes/sun/security/ssl/TrustManagerFactoryImpl.java~	2008-05-13 09:21:12.770538504 +0200
+++ openjdk/jdk/src/share/classes/sun/security/ssl/TrustManagerFactoryImpl.java	2008-05-13 09:24:55.533898839 +0200
@@ -146,7 +146,7 @@
                                 "javax.net.ssl.trustStoreProvider", ""));
                 props.put("trustStorePasswd", System.getProperty(
                                 "javax.net.ssl.trustStorePassword", ""));
-                props.put("/etc/pki/tls/certs/ca-bundle.crt", System.getProperty(
+                props.put("/etc/ssl/certs/ca-certificates.crt", System.getProperty(
                                 "javax.net.ssl.trustStore"));
 		props.put("CertBundle", System.getProperty(
                                 "javax.net.ssl.trustStoreType",
