# Don't set MAX_VM_MEMORY below MIN_VM_MEMORY

--- openjdk/jdk/make/docs/Makefile~
+++ openjdk/jdk/make/docs/Makefile
@@ -73,7 +73,7 @@
 else ifeq ($(ARCH),universal)
   MAX_VM_MEMORY = 1792
 else
-  MAX_VM_MEMORY = 768
+  MAX_VM_MEMORY = $(shell if [ $(MIN_VM_MEMORY) -gt 768 ]; then echo $(MIN_VM_MEMORY); else echo 768; fi)
 endif
 
 # List of all possible directories for javadoc to look for sources
