diff -ru openjdk-6-6b08.bak/jdk/src/solaris/bin/jexec.c openjdk-6-6b08.orig/jdk/src/solaris/bin/jexec.c
--- openjdk/jdk/src/solaris/bin/jexec.c	2008-03-26 09:59:57.000000000 +0100
+++ openjdk/jdk/src/solaris/bin/jexec.c	2008-04-03 17:14:10.000000000 +0200
@@ -164,9 +164,10 @@
 
     /* Get the path to the java binary, which is in a known position relative
      * to our current position, which is in argv[0]. */
-    if (getJavaPath(argv[argi++], java, RELATIVE_DEPTH) != 0) {
+    if (getJavaPath("/usr/lib/jvm/java-7-openjdk/jre/lib/jexec", java, RELATIVE_DEPTH) != 0) {
         errorExit(errno, MISSING_JAVA_MSG);
     }
+    argi++;
 
     nargv = (const char **) malloc((argc + 2) * (sizeof (const char *)));
     nargv[nargc++] = java;
