--- openjdk/jdk/make/common/shared/Compiler-gcc.gmk.orig	2009-12-20 23:22:47.000000000 +0100
+++ openjdk/jdk/make/common/shared/Compiler-gcc.gmk	2009-12-20 23:38:24.000000000 +0100
@@ -54,17 +54,17 @@
 ifeq ($(PLATFORM), linux)
 
   # Settings specific to Linux
-  CC             = $(COMPILER_PATH)gcc$(GCC_SUFFIX)
-  CPP            = $(COMPILER_PATH)gcc$(GCC_SUFFIX) -E
+  CC             = gcc$(GCC_SUFFIX)
+  CPP            = gcc$(GCC_SUFFIX) -E
   # statically link libstdc++ before C++ ABI is stablized on Linux
   STATIC_CXX     = true
   ifeq ($(STATIC_CXX),true)
     # g++ always dynamically links libstdc++, even we use "-Wl,-Bstatic -lstdc++"
     # We need to use gcc to statically link the C++ runtime. gcc and g++ use
     # the same subprocess to compile C++ files, so it is OK to build using gcc.
-    CXX            = $(COMPILER_PATH)gcc$(GCC_SUFFIX)
+    CXX            = gcc$(GCC_SUFFIX)
   else
-    CXX            = $(COMPILER_PATH)g++$(GCC_SUFFIX)
+    CXX            = g++$(GCC_SUFFIX)
   endif
   ifeq ($(ARCH), alpha)
     # alpha
