--- openjdk/corba/make/common/shared/Compiler-gcc.gmk~	2008-05-14 18:10:07.782578279 +0200
+++ openjdk/corba/make/common/shared/Compiler-gcc.gmk	2008-05-14 18:15:42.217623155 +0200
@@ -58,17 +58,17 @@
 ifeq ($(PLATFORM), linux)
 
   # Settings specific to Linux
-  CC             = $(COMPILER_PATH)gcc$(GCC_SUFFIX)
-  CPP            = $(COMPILER_PATH)gcc$(GCC_SUFFIX) -E
+  CC             = gcc$(GCC_SUFFIX)
+  CPP            = gcc$(GCC_SUFFIX) -E
   # statically link libstdc++ before C++ ABI is stablized on Linux
   STATIC_CXX     = true
   ifeq ($(STATIC_CXX),true)
     # g++ always dynamically links libstdc++, even we use "-Wl,-Bstatic -lstdc++"
     # We need to use gcc to statically link the C++ runtime. gcc and g++ use
     # the same subprocess to compile C++ files, so it is OK to build using gcc.
-    CXX            = $(COMPILER_PATH)gcc$(GCC_SUFFIX)
+    CXX            = gcc$(GCC_SUFFIX)
   else
-    CXX            = $(COMPILER_PATH)g++$(GCC_SUFFIX)
+    CXX            = g++$(GCC_SUFFIX)
   endif
   ifdef ICEDTEA_ZERO_BUILD
     # zero
--- openjdk/jdk/make/common/shared/Compiler-gcc.gmk~	2008-05-14 18:17:30.669259124 +0200
+++ openjdk/jdk/make/common/shared/Compiler-gcc.gmk	2008-05-14 18:58:41.196526465 +0200
@@ -58,17 +58,17 @@
 ifeq ($(PLATFORM), linux)
 
   # Settings specific to Linux
-  CC             = $(COMPILER_PATH)gcc$(GCC_SUFFIX)
-  CPP            = $(COMPILER_PATH)gcc$(GCC_SUFFIX) -E
+  CC             = gcc$(GCC_SUFFIX)
+  CPP            = gcc$(GCC_SUFFIX) -E
   # statically link libstdc++ before C++ ABI is stablized on Linux
   STATIC_CXX     = true
   ifeq ($(STATIC_CXX),true)
     # g++ always dynamically links libstdc++, even we use "-Wl,-Bstatic -lstdc++"
     # We need to use gcc to statically link the C++ runtime. gcc and g++ use
     # the same subprocess to compile C++ files, so it is OK to build using gcc.
-    CXX            = $(COMPILER_PATH)gcc$(GCC_SUFFIX)
+    CXX            = gcc$(GCC_SUFFIX)
   else
-    CXX            = $(COMPILER_PATH)g++$(GCC_SUFFIX)
+    CXX            = g++$(GCC_SUFFIX)
   endif
   ifdef ICEDTEA_ZERO_BUILD
     # zero
