--- openjdk/hotspot/src/share/vm/runtime/arguments.cpp~	2011-04-20 16:44:21.352657106 +0200
+++ openjdk/hotspot/src/share/vm/runtime/arguments.cpp	2011-04-20 16:45:33.702762924 +0200
@@ -59,7 +59,7 @@
 #include "gc_implementation/concurrentMarkSweep/compactibleFreeListSpace.hpp"
 #endif
 
-#define DEFAULT_VENDOR_URL_BUG "http://icedtea.classpath.org/bugzilla"
+#define DEFAULT_VENDOR_URL_BUG "https://bugs.launchpad.net/ubuntu/+source/openjdk-7/"
 #define DEFAULT_JAVA_LAUNCHER  "generic"
 
 char**  Arguments::_jvm_flags_array             = NULL;
