--- openjdk/hotspot/src/share/vm/runtime/arguments.cpp~	2008-03-04 11:25:12.915602117 +0000
+++ openjdk/hotspot/src/share/vm/runtime/arguments.cpp	2008-03-04 13:07:20.618037114 +0000
@@ -28,7 +28,7 @@
 #include "incls/_precompiled.incl"
 #include "incls/_arguments.cpp.incl"
 
-#define DEFAULT_VENDOR_URL_BUG "http://icedtea.classpath.org/bugzilla"
+#define DEFAULT_VENDOR_URL_BUG "https://bugs.launchpad.net/ubuntu/+source/openjdk-6/"
 #define DEFAULT_JAVA_LAUNCHER  "generic"
 
 char**  Arguments::_jvm_flags_array             = NULL;
--- openjdk/jdk/make/common/shared/Defs.gmk~	2008-03-22 20:49:41.239822027 +0100
+++ openjdk/jdk/make/common/shared/Defs.gmk	2008-03-22 20:51:20.331316799 +0100
@@ -194,13 +194,13 @@
 
 # Default names
 LAUNCHER_NAME = java
-PRODUCT_NAME = IcedTea
+PRODUCT_NAME = Java(TM)
 PRODUCT_SUFFIX = SE Runtime Environment
 COMPANY_NAME = Sun Microsystems, Inc.
 
 ifdef OPENJDK
   LAUNCHER_NAME = java
-  PRODUCT_NAME = IcedTea
+  PRODUCT_NAME = OpenJDK
   PRODUCT_SUFFIX = Runtime Environment
   COMPANY_NAME =
 endif
