--- openjdk/hotspot/src/os_cpu/linux_zero/vm/globals_linux_zero.hpp.orig	2013-12-19 13:53:15.987056600 +0000
+++ openjdk/hotspot/src/os_cpu/linux_zero/vm/globals_linux_zero.hpp	2013-12-19 13:58:01.041010426 +0000
@@ -34,7 +34,7 @@
 define_pd_global(bool,  DontYieldALot,           false);
 define_pd_global(intx,  ThreadStackSize,         1664);
 #ifdef _LP64
-define_pd_global(intx,  VMThreadStackSize,       1024);
+define_pd_global(intx,  VMThreadStackSize,       1664);
 #else
 define_pd_global(intx,  VMThreadStackSize,       512);
 #endif // _LP64
--- openjdk/corba/make/common/shared/Platform.gmk.orig	2013-12-19 13:51:15.766234650 +0000
+++ openjdk/corba/make/common/shared/Platform.gmk	2013-12-19 13:53:15.987056600 +0000
@@ -154,7 +154,7 @@
   # Arch and OS name/version
   mach := $(shell uname -m)
   ifneq (,$(wildcard /usr/bin/dpkg-architecture))
-    mach := $(shell (dpkg-architecture -qDEB_HOST_ARCH_CPU 2>/dev/null || echo $(mach)) | sed 's/arm64/aarch64/;s/powerpc$$/ppc/;s/hppa/parisc/')
+    mach := $(shell (dpkg-architecture -qDEB_HOST_ARCH_CPU 2>/dev/null || echo $(mach)) | sed 's/arm64/aarch64/;s/powerpc$$/ppc/;s/hppa/parisc/;s/ppc64el/ppc64le/')
   endif
   archExpr = case "$(mach)" in \
                 i[3-9]86) \
--- openjdk/jdk/make/jdk_generic_profile.sh.orig	2013-10-16 14:17:14.000000000 +0000
+++ openjdk/jdk/make/jdk_generic_profile.sh	2013-12-19 13:56:16.628292867 +0000
@@ -260,6 +260,7 @@
     sparc64) ZERO_LIBARCH=sparcv9   ;;
     arm*)    ZERO_LIBARCH=arm       ;;
     sh*)     ZERO_LIBARCH=sh        ;;
+    ppc64le) ZERO_LIBARCH=ppc64le   ;;
     *)       ZERO_LIBARCH="$(arch)"
   esac
   export ZERO_LIBARCH
@@ -269,7 +270,7 @@
     arm|i386|ppc|s390|sh|sparc)
       ARCH_DATA_MODEL=32
       ;;
-    aarch64|alpha|amd64|ia64|ppc64|s390x|sparcv9)
+    aarch64|alpha|amd64|ia64|ppc64*|s390x|sparcv9)
       ARCH_DATA_MODEL=64
       ;;
     *)
@@ -280,7 +281,7 @@
 
   # ZERO_ENDIANNESS is the endianness of the processor
   case "${ZERO_LIBARCH}" in
-    arm|aarch64|amd64|i386|ia64|mipsel)
+    arm|aarch64|amd64|i386|ia64|mipsel|ppc64le)
       ZERO_ENDIANNESS=little
       ;;
     ppc*|s390*|sparc*|alpha)
@@ -307,7 +308,7 @@
     s390)
       ZERO_ARCHFLAG="-m31"
       ;;
-    arm|aarch64)
+    arm|aarch64|ppc64le)
       ZERO_ARCHFLAG="-D_LITTLE_ENDIAN"
       ;;
     *)
--- openjdk/jdk/make/common/shared/Platform.gmk.orig	2013-12-19 13:51:15.754234564 +0000
+++ openjdk/jdk/make/common/shared/Platform.gmk	2013-12-19 13:53:15.987056600 +0000
@@ -161,7 +161,7 @@
     mach := $(shell uname -m)
   endif
   ifneq (,$(wildcard /usr/bin/dpkg-architecture))
-    mach := $(shell (dpkg-architecture -qDEB_HOST_ARCH_CPU 2>/dev/null || echo $(mach)) | sed 's/arm64/aarch64/;s/powerpc$$/ppc/;s/hppa/parisc/')
+    mach := $(shell (dpkg-architecture -qDEB_HOST_ARCH_CPU 2>/dev/null || echo $(mach)) | sed 's/arm64/aarch64/;s/powerpc$$/ppc/;s/hppa/parisc/;s/ppc64el/ppc64le/')
   endif
   archExpr = case "$(mach)" in \
                 i[3-9]86) \
@@ -205,10 +205,10 @@
     # Most archs are 32-bit
     ifndef ARCH_DATA_MODEL
       ARCH_DATA_MODEL=32
-      ifeq ($(ARCH), amd64)
+      ifneq (,$(findstring 64,$(ARCH)))
         ARCH_DATA_MODEL=64
       endif
-      ifeq ($(ARCH), ia64)
+      ifeq ($(ARCH), s390x)
         ARCH_DATA_MODEL=64
       endif
       ifeq ($(ARCH), sh)
