Index: openjdk/hotspot/src/share/vm/runtime/vmStructs.cpp
===================================================================
--- openjdk.orig/hotspot/src/share/vm/runtime/vmStructs.cpp	2011-07-10 13:43:01.082698001 +0200
+++ openjdk/hotspot/src/share/vm/runtime/vmStructs.cpp	2011-07-10 13:49:57.090698001 +0200
@@ -1281,7 +1281,6 @@
                                                                           \
   declare_toplevel_type(StubQueue)                                        \
   declare_toplevel_type(StubRoutines)                                     \
-  IA32_ONLY(declare_toplevel_type(StubRoutines::x86))                     \
   declare_toplevel_type(Stub)                                             \
            declare_type(InterpreterCodelet, Stub)                         \
                                                                           \
