--- openjdk/hotspot/src/os/solaris/launcher/java_md.c.orig	2008-11-20 09:41:00.000000000 +0100
+++ openjdk/hotspot/src/os/solaris/launcher/java_md.c	2008-12-02 15:07:35.000000000 +0100
@@ -1011,8 +1011,19 @@
         char buf[PATH_MAX+1];
         int len = readlink(self, buf, PATH_MAX);
         if (len >= 0) {
+	    const char* prefixes[] = {"/cow/", "/persistmnt/", "/rofs/", "/rwfs/", "/squashmnt/", NULL};
+	    const char **prefix;
+	    size_t prefix_len = 0;
             buf[len] = '\0';            /* readlink doesn't nul terminate */
-            exec_path = strdup(buf);
+	    for (prefix = prefixes; *prefix; prefix++) {
+	      prefix_len = strlen(buf) < strlen(*prefix) ? strlen(buf) : strlen(*prefix);
+	      if (!strncmp(*prefix, buf, prefix_len)) {
+		    prefix_len--;
+		    break;
+	      }
+	      prefix_len = 0;
+	    }
+	    exec_path = strdup(buf + prefix_len);
         }
     }
 #else /* !__sun && !__linux */
--- openjdk/hotspot/src/os/linux/launcher/java_md.c.orig	2008-12-02 15:10:43.000000000 +0100
+++ openjdk/hotspot/src/os/linux/launcher/java_md.c	2008-12-02 15:05:13.000000000 +0100
@@ -1011,8 +1011,19 @@
         char buf[PATH_MAX+1];
         int len = readlink(self, buf, PATH_MAX);
         if (len >= 0) {
+	    const char* prefixes[] = {"/cow/", "/persistmnt/", "/rofs/", "/rwfs/", "/squashmnt/", NULL};
+	    const char **prefix;
+	    size_t prefix_len = 0;
             buf[len] = '\0';            /* readlink doesn't nul terminate */
-            exec_path = strdup(buf);
+	    for (prefix = prefixes; *prefix; prefix++) {
+	      prefix_len = strlen(buf) < strlen(*prefix) ? strlen(buf) : strlen(*prefix);
+	      if (!strncmp(*prefix, buf, prefix_len)) {
+		    prefix_len--;
+		    break;
+	      }
+	      prefix_len = 0;
+	    }
+	    exec_path = strdup(buf + prefix_len);
         }
     }
 #else /* !__sun && !__linux */
--- openjdk/jdk/src/solaris/bin/java_md.c.orig	2008-11-20 09:44:52.000000000 +0100
+++ openjdk/jdk/src/solaris/bin/java_md.c	2008-12-02 15:02:54.000000000 +0100
@@ -868,8 +868,19 @@
         char buf[PATH_MAX+1];
         int len = readlink(self, buf, PATH_MAX);
         if (len >= 0) {
+	    const char* prefixes[] = {"/cow/", "/persistmnt/", "/rofs/", "/rwfs/", "/squashmnt/", NULL};
+	    const char **prefix;
+	    size_t prefix_len = 0;
             buf[len] = '\0';            /* readlink doesn't nul terminate */
-            exec_path = JLI_StringDup(buf);
+	    for (prefix = prefixes; *prefix; prefix++) {
+	      prefix_len = strlen(buf) < strlen(*prefix) ? strlen(buf) : strlen(*prefix);
+	      if (!strncmp(*prefix, buf, prefix_len)) {
+		    prefix_len--;
+		    break;
+	      }
+	      prefix_len = 0;
+	    }
+            exec_path = JLI_StringDup(buf + prefix_len);
         }
     }
 #else /* !__solaris__ && !__linux */
