--- openjdk/hotspot/make/linux/makefiles/gcc.make~	2012-12-01 18:49:19.202356557 +0100
+++ openjdk/hotspot/make/linux/makefiles/gcc.make	2012-12-01 18:50:41.430358414 +0100
@@ -84,7 +84,7 @@
 # Compiler flags
 
 # position-independent code
-ifneq ($(filter parisc ppc ppc64 s390 s390x sparc sparc64 sparcv9,$(ZERO_LIBARCH)),)
+ifneq ($(filter alpha m68k mips mipsel parisc ppc ppc64 s390 s390x sparc sparc64 sparcv9,$(ZERO_LIBARCH)),)
 PICFLAG = -fPIC
 else
 PICFLAG = -fpic
