Author: Andrew Haley <aph@redhat.com>
From: http://mail.openjdk.java.net/pipermail/distro-pkg-dev/2012-February/017196.html

--- openjdk/hotspot/make/linux/makefiles/defs.make.orig
+++ openjdk/hotspot/make/linux/makefiles/defs.make
@@ -132,7 +132,7 @@
   # Full Debug Symbols is supported on JDK7 or newer
 
   # Default OBJCOPY comes from GNU Binutils on Linux:
-  DEF_OBJCOPY=/usr/bin/objcopy
+  DEF_OBJCOPY=
   ifdef CROSS_COMPILE_ARCH
     # don't try to generate .debuginfo files when cross compiling
     _JUNK_ := $(shell \
@@ -163,7 +163,7 @@
     # Oracle security policy requires "all_strip". A waiver was granted on
     # 2011.09.01 that permits using "min_strip" in the Java JDK and Java JRE.
     #
-    DEF_STRIP_POLICY="min_strip"
+    DEF_STRIP_POLICY="no_strip"
     ifeq ($(ALT_STRIP_POLICY),)
       STRIP_POLICY=$(DEF_STRIP_POLICY)
     else
