diff -ruNE openjdk/hotspot/src/share/vm/interpreter/cppInterpreter.hpp openjdk-6-6b16~pre4/build/openjdk/hotspot/src/share/vm/interpreter/cppInterpreter.hpp
--- openjdk/hotspot/src/share/vm/interpreter/cppInterpreter.hpp	2009-04-30 23:21:47.000000000 +0000
+++ openjdk/hotspot/src/share/vm/interpreter/cppInterpreter.hpp	2009-06-29 12:30:31.000000000 +0000
@@ -60,8 +60,13 @@
 
 
   // No displatch table to switch so no need for these to do anything special
+#if defined(PRODUCT) && defined(HOTSPOT_ASM)
+  static void notice_safepoints();
+  static void ignore_safepoints();
+#else
   static void notice_safepoints() {}
   static void ignore_safepoints() {}
+#endif
 
   static address    native_result_to_tosca()                    { return (address)_native_abi_to_tosca; } // aka result handler
   static address    tosca_result_to_stack()                     { return (address)_tosca_to_stack; }
